### ISSUE TITLE
Customer Segmentation (Incomplete)

### ISSUE THREAD

**Issue by OLMENDY** [OLMENDY/-Incomplete-Customer-Segmentation#1]
Introduction
We will use clustering techniques to group customers based on their buying habits, using the sales dataset found on data. world. Customer segmentation allows businesses to understand the unique needs of each group and customize their marketing strategies and services accordingly. We will use K-means clustering, Hierarchical clustering, and DBSCAN, to segment the customers and compare the techniques to determine the best segmentation for our data.

Dataset
The dataset used for this project can be found at the following URL: SalesData on data.world
The dataset includes information about individual sales transactions, including the date of purchase, the customer's age and gender, the country and state of purchase, the product category and subcategory, the quantity sold, the unit cost and price, and the total cost and revenue of the transaction.

By analysing and segmenting this data, we hope to gain insights that will allow us to understand our customers better and serve their needs more effectively.

# %%
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from matplotlib import pyplot as plt

# %%
dataset = pd.read_excel("/Users/o.mendy/Documents/Learning/Projects/Segmentation/Sales_For_Course_quizz_table.xlsx")

# %%
dataset.head()

# %%
dataset.tail()

# %%
dataset = dataset.drop(columns='Column1')
dataset = dataset.dropna()

# %%
genders = dataset.value_counts('Customer Gender')
sns.set_style("whitegrid")
plt.figure(figsize=(10,4))
sns.barplot(x=genders.index, y=genders.values)
plt.show()

# %%
    dataset.tail()

# %%
Dataset_Demography = dataset.drop(columns = ['Date', 'Year', 'Month', 'State', 'Sub Category', 'Quantity', 'Unit Cost', 'Unit Price', 'Cost'], axis = 1)

# %%
Dataset_Demography.head()

# %%
Dataset_Demography.info()

# %%
Dataset_Demography.describe()

# %%
Dataset_Demography.isnull().sum()

# %%
Dataset_Demography.dropna(inplace=True)

# %%
Dataset_Demography.isnull().sum()

# %%
Dataset_Demography.describe()

# %%
Dataset_Demography.info()

# %%
Dataset_Demography.head()

# %%
from sklearn.cluster import KMeans
from sklearn.preprocessing import LabelEncoder

# %%
# Convert non-numeric variables into numeric ones
le = LabelEncoder()
Dataset_Demography['Customer Gender'] = le.fit_transform(Dataset_Demography['Customer Gender'])
Dataset_Demography['Country'] = le.fit_transform(Dataset_Demography['Country'])
Dataset_Demography['Product Category'] = le.fit_transform(Dataset_Demography['Product Category'])

# %%
# Determine the optimal number of clusters using Elbow Method
wcss = []  # within cluster sum of squares
for i in range(1, 11):
    kmeans = KMeans(n_clusters=i, init='k-means++', max_iter=300, n_init=10, random_state=0)
    kmeans.fit(Dataset_Demography)
    wcss.append(kmeans.inertia_)
    
plt.plot(range(1, 11), wcss)
plt.title('Elbow Method')
plt.xlabel('Number of clusters')
plt.ylabel('WCSS')
plt.show()

# %%
# Run K-Means with the optimal number of clusters determined by the Elbow Method
# For example, if the optimal number of clusters is 4
kmeans = KMeans(n_clusters=4, init='k-means++', max_iter=300, n_init=10, random_state=0)
clusters = kmeans.fit_predict(Dataset_Demography)

# %%
# Add the cluster column to the dataframe
Dataset_Demography['cluster'] = clusters

# %%
# Take a look at the dataset with the clusters
Dataset_Demography.head()

# %%
from scipy.cluster.hierarchy import linkage, dendrogram
from sklearn.preprocessing import MinMaxScaler

# %%
# Normalize the data
scaler = MinMaxScaler()
Dataset_Demography_scaled = pd.DataFrame(scaler.fit_transform(Dataset_Demography), columns = Dataset_Demography.columns)

# %%
# Create a linkage matrix using ward method
linkage_matrix = linkage(Dataset_Demography_scaled, 'ward')

# %%
# Plot a dendrogram using the linkage matrix
plt.figure(figsize=(10,7))
dendrogram(linkage_matrix)
plt.title('Hierarchical Clustering Dendrogram (Ward)')
plt.xlabel('Sample index')
plt.ylabel('Euclidean distances')
plt.show()

# %%
# From the dendrogram, decide on a number of clusters
# For example, we can choose 4 clusters, then we can cut the dendrogram:

from scipy.cluster.hierarchy import fcluster

# %%
# Cut the dendrogram at the height that will give us a certain number of clusters, e.g. 4
max_d = 25 # this is approximately the distance where we'd cut the dendrogram to get 4 clusters
clusters_h = fcluster(linkage_matrix, max_d, criterion='distance')

# %%
# Add the hierarchical clustering labels to the dataframe
Dataset_Demography['cluster_h'] = clusters_h

# %%
# Take a look at the dataset with the clusters
Dataset_Demography.head()

# %%
from sklearn.cluster import DBSCAN
from sklearn.metrics import silhouette_score

# %%
# DBSCAN
dbscan = DBSCAN(eps=0.5, min_samples=5)
clusters_d = dbscan.fit_predict(Dataset_Demography_scaled)

# %%
# Add the DBSCAN cluster labels to the dataframe
Dataset_Demography['cluster_d'] = clusters_d

# %%
# Take a look at the dataset with the clusters
Dataset_Demography.head()

# %%
# Now compare the silhouette scores of the three models
# Note that silhouette score requires the data to be in a certain format, so we're using the scaled data here
models = [kmeans, dbscan]
model_names = ['K-means', 'DBSCAN']

for model, name in zip(models, model_names):
    labels = model.labels_
    print(f'Silhouette score for {name}: {silhouette_score(Dataset_Demography_scaled, labels)}')

# Silhouette score for hierarchical clustering
print(f'Silhouette score for Hierarchical: {silhouette_score(Dataset_Demography_scaled, clusters_h)}')

# %%
# We can also visually compare the clusters from the different models
# For simplicity, let's just use two variables for plotting

# Plot the clusters from K-means
plt.figure(figsize=(10, 5))
plt.scatter(Dataset_Demography['Customer Age'], Dataset_Demography['Revenue'], c=Dataset_Demography['cluster'], cmap='viridis')
plt.title('K-means Clusters')
plt.xlabel('Customer Age')
plt.ylabel('Revenue')
plt.show()

# Plot the clusters from Hierarchical clustering
plt.figure(figsize=(10, 5))
plt.scatter(Dataset_Demography['Customer Age'], Dataset_Demography['Revenue'], c=Dataset_Demography['cluster_h'], cmap='viridis')
plt.title('Hierarchical Clusters')
plt.xlabel('Customer Age')
plt.ylabel('Revenue')
plt.show()

# Plot the clusters from DBSCAN
plt.figure(figsize=(10, 5))
plt.scatter(Dataset_Demography['Customer Age'], Dataset_Demography['Revenue'], c=Dataset_Demography['cluster_d'], cmap='viridis')
plt.title('DBSCAN Clusters')
plt.xlabel('Customer Age')
plt.ylabel('Revenue')
plt.show()

# %% 
from sklearn.metrics import silhouette_score

# For K-means
kmeans_silhouette = silhouette_score(Dataset_Demography_scaled, kmeans.labels_)
print("K-means silhouette score: ", kmeans_silhouette)

# For Agglomerative Clustering
agg_silhouette = silhouette_score(Dataset_Demography_scaled, agg_clustering.labels_)
print("Agglomerative Clustering silhouette score: ", agg_silhouette)

# For DBSCAN
dbscan_silhouette = silhouette_score(Dataset_Demography_scaled, dbscan.labels_)
print("DBSCAN silhouette score: ", dbscan_silhouette)
